pkg/controller/node/node_controller.go

```diff
@@ -0,0 +1,131 @@
+	// nodeControllerAgentName is the controller name for the Node controller
+	nodeControllerAgentName = "node"
+	// nodeEgressIPConfigAnnotationKey is the annotation key used for indicating the node's egress IP configuration
+	nodeEgressIPConfigAnnotationKey = "cloud.network.openshift.io/egress-ipconfig"
+)
+
+// NodeController is the controller implementation for Node resources
+// This controller is used to annotate nodes for the purposes of the
+// cloud network config controller
+type NodeController struct {
+	// Implements its own Node lister
+	nodesLister corelisters.NodeLister
+	// CloudProviderClient is a client interface allowing the controller
+	// access to the cloud API
+	CloudProviderClient cloudprovider.CloudProviderIntf
+	// KubeClientset is a standard kubernetes clientset
+	KubeClientset kubernetes.Interface
+}
+
+// NewNodeController returns a new Node controller
+func NewNodeController(
+	kubeClientset kubernetes.Interface,
+	cloudProviderClient cloudprovider.CloudProviderIntf,
+	nodeInformer coreinformers.NodeInformer) *controller.CloudNetworkConfigController {
+
+	nodeController := &NodeController{
+		nodesLister:         nodeInformer.Lister(),
+		KubeClientset:       kubeClientset,
+		CloudProviderClient: cloudProviderClient,
+	}
+
+	controller := controller.NewCloudNetworkConfigController(
+		[]cache.InformerSynced{nodeInformer.Informer().HasSynced},
+		nodeController,
+		nodeControllerAgentName,
+		nodeControllerAgentType,
+	)
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.Enqueue,
+	})
+	return controller
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Node resource
+// with the current status of the resource.
+func (n *NodeController) SyncHandler(key string) error {
+	// Convert the key to a name (since Node is non-namespaced)
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+	node, err := n.nodesLister.Get(name)
+	if err != nil {
+		// The Node resource may no longer exist, in which case we stop processing.
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("corev1.Node: '%s' in work queue no longer exists", key))
+			return nil
+		}
+		return fmt.Errorf("error retrieving corev1.Node from the API server, err: %v", err)
+	}
+	// If the node already has the annotation (ex: if we restart it is expected
+	// that the nodes would) we skip it. Subnets won't change and we are only
+	// interested in conveying the default assignment capacity that the node had
+	// when it started existing. It's up to the network plugin to track how much
+	// capacity it has left depending on the assignments it performs.
+	annotations := node.GetAnnotations()
+	if _, ok := annotations[nodeEgressIPConfigAnnotationKey]; ok {
```

Comment:
The getters are part of the `metav1.Object` interface, so that in theory you can have APIs that work on any API object type, but this almost never happens in practice (at least, outside of the apiserver).